### ISSUE TITLE
Can't load .ufo file(folders) on macOS

### ISSUE THREAD

**Issue by hskim08** [trufont/trufont#681]
.ufo shows up as a folder on macOS. However, the `FileMode` for `QFileDialog` is set to `QFileDialog.ExistingFile` which does not allow loading folders. (`trufont/Lib/trufont/objects/application.py line 444)

Using the default (`QFileDialog.AnyFile`) file mode will fix this issue. However this puts the onus on the user to open the correct file format.
I would be happy to submit a pull request if needed.

**Comment by madig** [comment]
If you find some nice Qt mode for doing the right thing, I'd appreciate a PR. I think macOS and non-macOS need different handling here because a macOS package is seen as a file...

**Comment by hskim08** [comment]
On macOS Mojave(10.14.6) and Catalina (10.15.1) .ufo shows up as a folder.
(Unless I'm missing some UFO related install package)
In that case it seems reasonable to pull `ufoFormat` out of `if packageAsFile:`.

**Comment by madig** [comment]
Applications on macOS are able to open UFOs like you'd open a file, no installation of anything necessary. Will have a look at this when I get home today or this weekend.

**Comment by madig** [comment]
I can double-click on UFOs in TruFont on macOS just like a file. Maybe there's something else going on.

**Comment by hskim08** [comment]
Here's a .ufo file I created with TruFont on my Desktop. It shows up as a folder.
![Screen Shot 2019-11-02 at 5 53 05 PM](https://user-images.githubusercontent.com/721281/68078732-c517b680-fd99-11e9-956e-5dc1ce90185f.png)

When I try to open it in TruFont, I can't because it's a folder.
![Screen Shot 2019-11-02 at 5 55 42 PM](https://user-images.githubusercontent.com/721281/68078751-05773480-fd9a-11e9-8500-e22058e1362e.png)

Did you set up something on your mac to make the OS view .ufo folders as a package/bundle instead of a folder? To my knowledge, for a folder to be viewed as a bundle it needs to have an Info.plist (https://apple.stackexchange.com/questions/280280/what-makes-an-macos-package-show-as-a-package)

Can you check your .ufo file to see if it has an Info.plist?

**Comment by madig** [comment]
Nope, no Info.plist. I have Glyphs installed however, maybe it registers something?

**Comment by hskim08** [comment]
This is getting interesting.
I downloaded Glyphs and ran the app which changed the .ufo folder into a package. 
It looks like the `Info.plist` of the app, not the package, needs to specify that .ufo is a package and that the app can open it. (The block below is from the Info.plist from Glyphs.)
I will need to take a further look into how the app is bundled for macOS and see if there is a way to modify the `Info.plist` for TruFont.

```
<key>CFBundleDocumentTypes</key>
<array>
	<dict>
		<key>CFBundleTypeName</key>
		<string>UnifiedFontObject</string>
		<key>CFBundleTypeRole</key>
		<string>Editor</string>
		<key>LSItemContentTypes</key>
		<array>
			<string>org.unifiedfontobject.ufo</string>
		</array>
		<key>LSTypeIsPackage</key>
		<true/>
		<key>NSDocumentClass</key>
		<string>GSDocument</string>
		<key>NSExportableTypes</key>
		<array>
			<string>org.unifiedfontobject.ufo</string>
		</array>
	</dict>
</array>
<key>UTImportedTypeDeclarations</key>
<array>
	<dict>
		<key>UTTypeConformsTo</key>
		<array>
			<string>com.apple.package</string>
		</array>
		<key>UTTypeDescription</key>
		<string>Unified Font Object</string>
		<key>UTTypeIdentifier</key>
		<string>org.unifiedfontobject.ufo</string>
		<key>UTTypeReferenceURL</key>
		<string>https://unifiedfontobject.org</string>
		<key>UTTypeTagSpecification</key>
		<dict>
			<key>public.filename-extension</key>
			<array>
				<string>ufo</string>
			</array>
		</dict>
	</dict>
</array>
```

**Comment by hskim08** [comment]
As an added note, the following command will reset the Uniform Type Identifiers of the OS:
```
/System/Library/Frameworks/CoreServices.framework/Versions/A/Frameworks/LaunchServices.framework/Versions/A/Support/lsregister -kill -seed
```
(Lifted from https://apple.stackexchange.com/questions/329910/reset-uniform-type-identifiers-on-mac-os-x?rq=1)

**Comment by madig** [comment]
I suppose this means that TruFont would have to be properly packaged for macOS instead of just being some Python package somewhere.

**Comment by hskim08** [comment]
Yes, it seems so. 
I added the code block above to `Info.plist` for the released version for macOS (vers. 0.4.0)
and after logging off then on, the OS changed the .ufo folder to a package and TruFont was set as the default app for opening the .ufo file.

**Comment by madig** [comment]
I think @adrientetar rolled them manually...

**Comment by spl** [comment]
I was just trying out TruFont and ran into the issue posted at the top. And, indeed, changing `QFileDialog.ExistingFile` to `QFileDialog.AnyFile` in my virtual environment's `lib/python3.8/site-packages/trufont/objects/application.py` allowed me to open a `.ufo` directory. However, if I ”opened” any other folder by either hitting “Return” ↩️ or clicking the “Open” button, TruFont crashed with: `[Errno 21] Is a directory: '...'`.

**Comment by madig** [comment]
PR welcome, I don't have other time for this project.

**Comment by hskim08** [comment]
I've started to look into this again.

First a quick workaround if you want to use TruFont immediately.
As a stop gap I have been using the following hack:
1. Download TruFont.app 0.5.0 - the version bundled with an app.
2. Right click and click "Show Package Contents"
3. Open Info.plist and copy the big chuck of XML from my comment above into between <dict></dict>.
4. Save the file.
5. Open Trufont.app once. This will register .ufo as a package into the system.

Once .ufo is registered, it will stay that way. Now you can run the latest TruFont from python and it will see .ufo as a package.

Now for a seemingly correct way that I found a moment ago.
It seems it is possible to set UTI handlers from python via pyobjc.
Here's a library that wraps the functionality in a nice class: <https://github.com/impact27/applaunchservices>
I'd like to incorporate this into TruFont, except for some reason I can't get macOS to unsee .ufo as a package.
So work in progress...

**Comment by hskim08** [comment]
I found some time to look into this further and I would like to share my findings and then discuss different possible solutions to this issue and what the path forward should be.

First, there are two parts to this issue:
1. Adding the UFO extension as a Uniform Type Identifier(UTI) to the launch service registry.
2. Letting the OS know to launch TruFont when it sees a UFO file.

It turns out `applaunchservices` handles the second part and not the first. So that will not fix this problem.

The main issue is the first one. This apparently is only possible through adding the information about UFO in the Info.plist in an app bundle. I have tried to look for a way to just register UFO to the launch service registry but it seems that is not how macOS handles UTIs.
So we are in a situation where if the user already installed and ran another font editor that handles UFO files, then everything will work as it is now but if TruFont is the first UFO application on the computer, we run into the situation where the OS thinks a UFO bundle is just a folder.

It seems we need to discuss what would be a good path forward, especially for new users (on macOS) i.e. people to whom TruFont is their first font editor.

### Solution 1. Let users pick the folder
This is similar to the original approach to this issue. If UFO isn't registered, we could let the user open folders. This requires the least code modification but puts the onus of choosing the right UFO folder.

### Solution 2. Create an Installer (via fman build system?) with correct Info.plist
This is probably the correct way forward for most users. This however wouldn't work if TruFont is run by command line. Also this opens up the question on what the path forward for TruFont is, a pypi package or a standalone application?
This would also change the scope of this issue, since now it will be about building installers.
I would be interested to go down this path in the name of better user experience.

### Solution 3. Let TruFont Decide
Through LaunchServices (via pyobjc) it is possible to query whether there are other applications that can handle a UTI. TruFont could do a quick UFO check on launch. If there are no UFO applications let the user open `metainfo.plist` like on linux (and windows?). If it is there, let the user choose the .ufo package.

This solution requires a quick check which would be less disruptive to implement. The check could be in `treatPackageAsFile`.
The drawback of this solution is that the file opening experience will be inconsistent depending on whether UFO is registered or not. 

### Solution 4. Run a dummy app with Info.plist that specifies UFO.
This is now completely hacky and may have problems if used with an installer, but we could create a UFO register app and run that once.

At this point, properly building installers seems the right point. But I am open for discussion before moving forward.
